### ISSUE TITLE
BUG: --lf running only a subset of last-failed tests

### ISSUE THREAD

**Issue by jbrockmendel** [pytest-dev/pytest#7758]
After running the pandas test suite and getting e.g. 20 test failures, running the same command with " --lf" appended should run those 20 tests.  In practice I'm seeing only fraction of these run on the subsequent run.

```
$ pytest pandas/tests --skip-slow --skip-db
[...]
=================================================================== short test summary info ====================================================================
FAILED pandas/tests/resample/test_resampler_grouper.py::test_apply_columns_multilevel - KeyError: Timestamp('2017-01-01 00:00:00', freq='H')
FAILED pandas/tests/groupby/test_function.py::test_median_empty_bins[True] - AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="0") are different
FAILED pandas/tests/groupby/test_function.py::test_ops_general[count-size] - AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="0") are different
FAILED pandas/tests/groupby/test_grouping.py::TestGrouping::test_evaluate_with_empty_groups[agg-expected1] - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_aggregate.py::test_aggregate_item_by_item - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_aggregate.py::test_multiindex_custom_func[<lambda>0] - AssertionError: Attributes of DataFrame.iloc[:, 0] (column ...
FAILED pandas/tests/groupby/aggregate/test_aggregate.py::test_multiindex_custom_func[<lambda>1] - AssertionError: Attributes of DataFrame.iloc[:, 0] (column ...
FAILED pandas/tests/groupby/aggregate/test_aggregate.py::test_multiindex_custom_func[<lambda>2] - AssertionError: Attributes of DataFrame.iloc[:, 0] (column ...
FAILED pandas/tests/groupby/aggregate/test_cython.py::test_cython_agg_empty_buckets[True-mean-mean] - AssertionError: Attributes of DataFrame.iloc[:, 0] (col...
FAILED pandas/tests/groupby/aggregate/test_cython.py::test_cython_agg_empty_buckets[True-median-<lambda>] - AssertionError: Attributes of DataFrame.iloc[:, 0...
FAILED pandas/tests/groupby/aggregate/test_cython.py::test_cython_agg_empty_buckets[True-var-<lambda>] - AssertionError: Attributes of DataFrame.iloc[:, 0] (...
FAILED pandas/tests/groupby/aggregate/test_other.py::test_agg_lambda_with_timezone - ValueError: Must produce aggregated value
FAILED pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[NotImplementedError] - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[RuntimeError] - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[KeyError] - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[IndexError] - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[OSError] - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[ValueError] - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[ArithmeticError] - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[AttributeError] - AssertionError: Attributes of Series are different
====================================== 20 failed, 5034 passed, 1652 skipped, 374 xfailed, 1 warning in 208.51s (0:03:28) =======================================

$ pytest pandas/tests/resample pandas/tests/groupby pandas/tests/generic/ --skip-slow --lf
[...]
============================================================== 1 failed, 1455 deselected in 0.68s ==============================================================
```

Running pytest 6.0.2, py3.8.5, OSX 10.15.5

**Comment by The-Compiler** [comment]
Can you show the `.pytest_cache/v/cache/lastfailed` file after the first run?

**Comment by jbrockmendel** [comment]
```
$ cat pytest_cache-bak/v/cache/lastfailed
{
  "pandas/tests/extension/decimal/test_decimal.py::test_groupby_agg": true,
  "pandas/tests/extension/decimal/test_decimal.py::test_groupby_agg_ea_method": true,
  "pandas/tests/groupby/aggregate/test_aggregate.py::test_aggregate_item_by_item": true,
  "pandas/tests/groupby/aggregate/test_aggregate.py::test_multiindex_custom_func[<lambda>0]": true,
  "pandas/tests/groupby/aggregate/test_aggregate.py::test_multiindex_custom_func[<lambda>1]": true,
  "pandas/tests/groupby/aggregate/test_aggregate.py::test_multiindex_custom_func[<lambda>2]": true,
  "pandas/tests/groupby/aggregate/test_cython.py::test_cython_agg_empty_buckets[True-mean-mean]": true,
  "pandas/tests/groupby/aggregate/test_cython.py::test_cython_agg_empty_buckets[True-median-<lambda>]": true,
  "pandas/tests/groupby/aggregate/test_cython.py::test_cython_agg_empty_buckets[True-var-<lambda>]": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_agg_lambda_with_timezone": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[ArithmeticError]": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[AttributeError]": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[IndexError]": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[KeyError]": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[NotImplementedError]": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[OSError]": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[RuntimeError]": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[ValueError]": true,
  "pandas/tests/groupby/test_function.py::test_median_empty_bins[True]": true,
  "pandas/tests/groupby/test_function.py::test_ops_general[count-size]": true,
  "pandas/tests/groupby/test_grouping.py::TestGrouping::test_evaluate_with_empty_groups[agg-expected1]": true,
  "pandas/tests/resample/test_resampler_grouper.py::test_apply_columns_multilevel": true
}%
```

The output from the second run:

```
========================================================================= short test summary info ==========================================================================
FAILED pandas/tests/extension/decimal/test_decimal.py::test_groupby_agg - AssertionError: Attributes of Series are different
FAILED pandas/tests/extension/decimal/test_decimal.py::test_groupby_agg_ea_method - AssertionError: Attributes of Series are different
============================================================= 2 failed, 30232 deselected, 2 warnings in 9.07s ==============================================================

```

**Comment by jbrockmendel** [comment]
Possibly relevant: im often running multiple pytest processes in separate directories

**Comment by nicoddemus** [comment]
> Possibly relevant: im often running multiple pytest processes in separate directories

At the same time? During each run, the `rootdir` on the pytest header is the same?

The cache mechanism used by `--lf` is not multi-process safe, so this can definitely be the reason if you are running long test suites at the same time.

**Comment by jbrockmendel** [comment]
> During each run, the rootdir on the pytest header is the same?

how do i determine the rootdir?

I have different branches in different git worktree directories, am only running one process at a time for any single branch, always with the cwd in that branch's directory

**Comment by nicoddemus** [comment]
> how do i determine the rootdir?

From your answer I assume they are different? 😉 You can use `--rootdir`, or even better, set the [`cache_dir`](https://docs.pytest.org/en/stable/reference.html?highlight=cache_dir#confval-cache_dir) configuration option, which will set the cache directory (used by `--lf`) explicitly.

> am only running one process at a time for any single branch,

Ahh OK thanks, so it is not a race condition.

**Comment by ojii** [comment]
I'm seeing the same issue:

```
$ pytest
[...]
4 failed, 1857 passed, 144 skipped
$ pytest --last-failed
[...]
run-last-failure: rerun previous 1 failure first (skipped 24 files)
[...]
1 failed, 870 deselected
```

is it incorrectly skipping too many files?

**Comment by ornichola** [comment]
I'm facing the same issue. My lastfailed file contains **34 items**:
```
(venv) user@macbook % pytest --cache-show
...
cache/lastfailed contains:
  {
  ...
  }
```

But running last failed tests select to run only **12 items**:
```
(venv) user@macbook % pytest --lf --collect-only
...
collected 53 items / 41 deselected / 12 selected
...
run-last-failure: rerun previous 12 failures (skipped 55 files)
```

**Comment by The-Compiler** [comment]
Perhaps something else (hook or plugin) is deselecting them? @ojii @jbrockmendel what plugins do you have installed? Does someone have a minimal, isolated example?

**Comment by jbrockmendel** [comment]
> what plugins do you have installed?

looks like hypothesis, pytest-xdist, pytest-cov, pytest-asyncio, pytest-forked

Some of these I can't do without, but I can try trimming a few and see if that affects anything.

**Comment by fergushev** [comment]
I'm seeing this as well and the problem (at least in my case) appears to be related to having an `__init__.py` file in the test directories. 

My setup looks something like this:
```
tests/functional/foo/  10 tests (dir has `__init__.py`)
tests/functional/bar/  10 tests (dir does _not_ have `__init__.py`)
tests/functional/baz/  10 tests (dir has `__init__.py`)
```
My `lastfailed` contains all 30 tests from the above directories. 
When I run with `--lf` it only picks up the tests from `tests/functional/bar/`. 
If I delete `__init__.py` from `/foo/` then `--lf` picks up 20 tests. 
If I delete `__init__.py` from `/baz/` then `--lf` picks up all 30 tests.

---

```
{'Python': '3.8.5', 'Platform': 'macOS-10.15.7-x86_64-i386-64bit', 
'Packages': {'pytest': '6.1.1', 'py': '1.9.0', 'pluggy': '0.13.1'}, 
'Plugins': {'metadata': '1.10.0', 'html': '1.20.0', 'xdist': '2.1.0', 'forked': '1.3.0'}}
```

**Comment by bluetech** [comment]
Pretty sure I saw this too, but wasn't in a situation where I could look into it. We should definitely look at what's going wrong here. In the mean time any sort of standalone/consistent reproduction steps would be very helpful.

**Comment by bluetech** [comment]
In #7894 @deb17 has provided some more details on this issue.

**Comment by bluetech** [comment]
Bisected @deb17's test case to 1b30514783d8ed4225a59af3f9f3fa668ab56f67 (pr #6448). So this is not a new issue, it regressed in 5.4.0. I think what happens is that the *package* containing a failing tests gets incorrectly filtered out. I'll try to figure out what goes wrong here.

**Comment by nicoddemus** [comment]
The problems with `--lf` all started with a very simple implementation from me that tried to speed up `--lf`, and that proved not to be sufficient, so we kept adding more and more complexity to it to the point where the code is really hard to follow/understand and contains subtle bugs.

I'm OK with just dropping all the extra complexity in there, it will make for slightly slower code, but simpler and more importantly, correct.
